rsa.d

```diff
@@ -0,0 +1,491 @@
+		size_t maxHeaderL	= 2 + 2 + 4;
+		size_t maxEKL		= EVP_PKEY_size(keypair);
+		size_t maxIVL		= EVP_MAX_IV_LENGTH;
+
+		ubyte* ek			= encMessage.ptr + maxHeaderL;
+		ubyte[8] temp		= 0;
+		temp[0..2]			= encMessage[0..2];
+		int ekl				= (cast(int[])temp)[0];
+			
+		ubyte* iv			= encMessage.ptr + maxHeaderL + maxEKL;
+		temp				= 0;
+		temp[0..2]			= encMessage[2..4];
+		size_t ivl			= (cast(int[])temp)[0];
+
+		ubyte* encMsg		= encMessage.ptr + maxHeaderL + maxEKL + maxIVL;
+		temp				= 0;
+		temp[0..4]			= encMessage[4..8];
+		size_t encMsgLen	= (cast(size_t[])temp)[0];
+		
+		size_t decLen   = 0;
+		size_t blockLen = 0;
+		EVP_PKEY *key;
+		ubyte* _decMsg;
+		auto decMsg = &_decMsg;
+		*decMsg = cast(ubyte*)GC.malloc(encMsgLen + ivl);
+		if(decMsg == null)
+			throw new CryptographicException("Malloc failed.");
+
+		EVP_CIPHER_CTX *rsaDecryptCtx = cast(EVP_CIPHER_CTX*)GC.malloc(EVP_CIPHER_CTX.sizeof);	
+		if(rsaDecryptCtx == null)
+			throw new CryptographicException("Malloc failed.");
+
+		EVP_CIPHER_CTX_init(rsaDecryptCtx);
+
+		if(!EVP_OpenInit(rsaDecryptCtx, EVP_aes_256_ctr(), ek, ekl, iv, keypair))
+			throw new CryptographicException("EVP_OpenInit failed.");
+
+		if(!EVP_OpenUpdate(rsaDecryptCtx, cast(ubyte*)*decMsg + decLen, cast(int*)&blockLen, encMsg, cast(int)encMsgLen))
+			throw new CryptographicException("EVP_OpenUpdate failed.");
+		decLen += blockLen;
+
+		if(!EVP_OpenFinal(rsaDecryptCtx, cast(ubyte*)*decMsg + decLen, cast(int*)&blockLen))
+			throw new CryptographicException("EVP_OpenFinal failed.");
+		decLen += blockLen;
+
+		EVP_CIPHER_CTX_cleanup(rsaDecryptCtx);
+
+		return (*decMsg)[0 .. decLen];
+	} // open()
+
+// ----------------------------------------------------------
+
+	ubyte[] getPublicKey()
+	{
+		BIO* bio = BIO_new(BIO_s_mem());
+
+		PEM_write_bio_PUBKEY(bio, keypair);
+
+		ubyte[] buffer = new ubyte[BIO_ctrl_pending(bio)];
+		BIO_read(bio, buffer.ptr, cast(int)buffer.length);
+		BIO_free_all(bio);
+
+		return buffer;
+	} // getPublicKey()
+
+// ----------------------------------------------------------
+
+	public ubyte[] getPrivateKey(string password, int iterations = 25000, bool use3Des = false)
+	{
+		if (!_hasPrivateKey)
+			return null;
+
+		BIO* bio = BIO_new(BIO_s_mem());
+
+		if (password is null)
+			PEM_write_bio_PKCS8PrivateKey(bio, keypair, null, null, 0, null, null);
+		else
+		{
+			ubyte[] pwd = cast(ubyte[])password;
+			pwd = pwd ~ '\0';
+
+			PEM_write_bio_PKCS8PrivateKey(
+				bio,
+				keypair,
+				!use3Des ? EVP_aes_256_cbc() : EVP_des_ede3_cbc(),
+				null,
+				0,
+				null,
+				pwd.ptr);
+		}
+
+		if(BIO_ctrl_pending(bio) == 0)
+			throw new CryptographicException("No private key written.");
+
+		ubyte[] buffer = new ubyte[BIO_ctrl_pending(bio)];
+		BIO_read(bio, buffer.ptr, cast(int)buffer.length);
+		BIO_free_all(bio);
+
+		return buffer;
+	} // getPrivateKey()
+
+// ----------------------------------------------------------
+
+	ubyte[] encrypt(const ubyte[] inMessage)
+	{
```

Comment:
RSA can only encrypt data that is less than or equal to it's key length, so in this case 2048/8=256 bytes. However, the amount of space available varies by the padding scheme used. Since you selected OEAP (The best available) this means that we must subtract 42 bytes of available space for padding overhead. This means that there is only 214 bytes available. This method should use an in { } constraint to verify that the message is less than or equal to 214 bytes. In the case of 4096 bits key length (512 bytes) it would be 480 bytes.